src/ranged.cpp

```diff
@@ -156,24 +156,74 @@ int player::gun_engagement_moves( const item &gun, int target, int start ) const
 }
+bool player::handle_gun_damage( item &it )
 
-bool player::handle_gun_damage( item &it, int shots_fired )
 {
     if( !it.is_gun() ) {
         debugmsg( "Tried to handle_gun_damage of a non-gun %s", it.tname() );
         return false;
     }
-
+    int dirt = it.get_var( "dirt", 0 );
+    int last_fired = it.get_var( "last_fired", 0 );
+    int just_unclogged = it.get_var( "just_unclogged", 0 );
     const auto &curammo_effects = it.ammo_effects();
     const cata::optional<islot_gun> &firing = it.type->gun;
+    const std::string thistimefired = to_string_time_of_day( calendar::turn );
+    // Here we check for a chance for the weapon to suffer a misfire due to
+    // being dirty (blackpowder fouling effect only; typical dirt accumulation
+    // from smokeless powder is not currently modeled by this effect).
+    // Can damage gun, but not enough to push the weapon beyond 'shattered'.
+    // Check, but not accumulation, will happen even when firing normal rounds,
+    // so you can screw up your MP5 by firing blackpowder rounds and the
+    // effects will be in force even when firing retail ammo, though the gun
+    // won't get MORE foul.
+
+    // causes failure to cycle if weapon is fired too quickly:
+    if( last_fired == to_turn<int>( calendar::turn ) && !it.has_flag( "PUMP_ACTION" ) &&
+        !it.has_flag( "MANUAL_ACTION" ) && it.type->gun->ups_charges < 1 &&
+        ( curammo_effects.count( "MUZZLE_SMOKE" ) || curammo_effects.count( "BLACKPOWDER" ) ) ) {
+        add_msg_player_or_npc( _( "Your %s fails to cycle!" ),
+                               _( "<npcname>'s %s fails to cycle!" ),
+                               it.tname() );
+        return false;
+    }
+    it.set_var( "last_fired", to_turn<int>( calendar::turn ) );
+    // chance to cause malfunction, which always happens when clogged:
+    int malfunctionreduction = 0;
+    if( it.has_flag( "PUMP_ACTION" ) || it.has_flag( "MANUAL_ACTION" ) ) {
+        malfunctionreduction = 30;
+    }
+    if( it.type->gun->ammo.count( ammotype( "flintlock" ) ) ||
+        ( it.type->gun->ammo.count( ammotype( "blunderbuss" ) ) ) ||
+        ( it.type->gun->ammo.count( ammotype( "36paper" ) ) ) ||
+        ( it.type->gun->ammo.count( ammotype( "44paper" ) ) ) ) {
+        malfunctionreduction = 39;
+    }
+    if( ( just_unclogged != 1 && dirt > 100 &&
+          one_in( ( 600 - dirt ) / ( 40 - malfunctionreduction ) ) ) ||
+        it.has_fault( fault_gun_clogged ) ) {
+        it.faults.insert( fault_gun_clogged );
+        it.faults.erase( fault_gun_blackpowder );
+        add_msg_player_or_npc( _( "Your foul %s misfires with a muffled click!" ),
+                               _( "<npcname>'s foul %s misfires with a muffled click!" ),
+                               it.tname() );
+        return false;
+    }
+    // chance to damage gun:
+    if( it.damage() < it.max_damage() &&
+        ( curammo_effects.count( "MUZZLE_SMOKE" ) || curammo_effects.count( "BLACKPOWDER" ) ) &&
```

Comment:
Might be better to check for BLACKPOWDER only, to potentially support something that creates smoke clouds if fired, but no fouling because of it.

Reply:
Not all BP ammo has the BLACKPOWDER flag. I guess I can go into JSON and add it.